### PR TITLE
glib: Add _once variants to many methods

### LINKED ISSUE #341: glib::timeout_add_local accept FnOnce
I sometimes find myself in the situation of adding a timeout/callback that will be called exactly once. But the respective `glib` methods accept a `FnMut` at best. One could wrap a `FnOnce` like this and expose this as an alternative (untested code):

```rust
fn wrap_once_local(func: F) -> impl (FnMut() -> Continue + 'static)
where
    F: FnOnce() 'static,  
{
    let mut func = Some(func);
    move || {
        func.take().unwrap();
        Continue(false)
    }
}
```

### PR DESCRIPTION
**Pull request by piegamesde**
Closes #341. Please don't mind the copy-paste blueprints in the code ^^

CC @sdroege

### CODE REVIEW

File: glib/src/source.rs [hunk 1]
```diff
@@ -142,6 +142,56 @@ fn into_raw_unix_fd<F: FnMut(RawFd, IOCondition) -> Continue + 'static>(func: F)
     Box::into_raw(func) as gpointer
 }
 
+///
+/// In comparisong to `timeout_add()`, this only requires `func` to be
+/// `FnOnce`, and will automatically return `Continue(false)`.
+where
+    F: FnOnce() + Send + 'static,
```
Comment (sdroege) [inline glib/src/source.rs:149]: Something seemed to have gone missing here

File: glib/src/source.rs [hunk 2]
```diff
@@ -142,6 +142,56 @@ fn into_raw_unix_fd<F: FnMut(RawFd, IOCondition) -> Continue + 'static>(func: F)
...
+/// in order to prevent being called twice.
+#[inline(always)]
+fn fnmut_callback_wrapper(func: F) -> G
+where
+    F: FnOnce() + Send + 'static,
+    G: FnMut() -> Continue + Send + 'static,
+{
+    let mut func = Some(func);
+    move || {
+        func.take().unwrap();
```
Comment (sdroege) [inline glib/src/source.rs:164]:
```suggestion
        func.take().expect("GSource closure called after returning glib::Continue(false)");
```

Or something like that

File: glib/src/source.rs [hunk 3]
```diff
@@ -142,6 +142,56 @@ fn into_raw_unix_fd<F: FnMut(RawFd, IOCondition) -> Continue + 'static>(func: F)
...
+        Continue(false)
+    }
+}
+
+///
+/// In comparisong to `timeout_add()`, this only requires `func` to be
```
Comment (sdroege) [inline glib/src/source.rs:170]:
```suggestion
/// In comparison to `timeout_add()`, this only requires `func` to be
```

File: glib/src/source.rs [hunk 4]
```diff
@@ -142,6 +142,56 @@ fn into_raw_unix_fd<F: FnMut(RawFd, IOCondition) -> Continue + 'static>(func: F)
...
+/// `FnOnce`, and will automatically return `Continue(false)`.
+where
+    F: FnOnce() + 'static,
```
Comment (sdroege) [inline glib/src/source.rs:173]: Also here something went missing

**Review by sdroege — COMMENTED** [review]
Otherwise the approach looks good, just need to get it to compile :)

File: glib/src/source.rs (hunk 1 above)
Comment (piegamesde) [reply glib/src/source.rs:149]: No, that's my copy-paste blueprint that I accidentally commited as well.

**Comment by piegamesde** [comment]
I had to replace the generics with `impl Trait` types, otherwise the compiler didn't infer the types correctly for some reason.

**Comment by sdroege** [comment]
> I had to replace the generics with `impl Trait` types, otherwise the compiler didn't infer the types correctly for some reason.

That's strange. This should be exactly the same apart from not having the generic types not nameable. But this way is fine too, all the same!

**Review by sdroege — APPROVED** [review]
Looks good to me, can you squash the commits?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by piegamesde** [comment]
Done.

**Pull request merged**